contractsV5/FeeHandler.sol

```diff
@@ -0,0 +1,102 @@
+        uint epoch = (brrAndEpochData / (1 << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        uint rewardInBPS = (brrAndEpochData / (1 << BITS_PER_PARAM << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        uint burnInBPS = (brrAndEpochData / (1 << BITS_PER_PARAM << BITS_PER_PARAM << BITS_PER_PARAM)) & (1 << BITS_PER_PARAM) - 1;
+        return (burnInBPS, rewardInBPS, epoch, expiryBlock);
+    }
+
+    // Todo: future optimisation to accumulate rebates for 2 rebate wallet
+    // encode totals, 128 bits per reward / rebate
+    function handleFees(address[] calldata eligibleWallets, uint[] calldata rebatePercentages) external payable returns(bool) {
+        // Decoding BRR data
+        (uint burnInBPS, uint rewardInBPS, uint epoch, uint expiryBlock) = decodeData();
+        uint rebateInBPS = BPS - rewardInBPS - burnInBPS;
+
+        // Check current block number
+        if(block.number > expiryBlock) {
+            (burnInBPS, rewardInBPS, rebateInBPS, epoch, expiryBlock) = kyberDAOContract.getLatestBRRData();
+
+            // Update brrAndEpochData
+            brrAndEpochData = encodeData(burnInBPS, rewardInBPS, epoch, expiryBlock);
+        }
+
+        uint rebateWei = rebateInBPS * msg.value / BPS;
+        uint rewardWei = rewardInBPS * msg.value / BPS;
+        for(uint i = 0; i < eligibleWallets.length; i ++) {
+            // Internal accounting for rebates per reserve wallet (totalRebatesPerRebateWallet)
+            totalRebatesPerRebateWallet[eligibleWallets[i]] = rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for total rebates (totalRebates)
+            totalRebates += rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for reward per epoch (totalRewardsPerEpoch)
+            totalRewardsPerEpoch[epoch] += rewardWei;
+            // Internal accounting for total rewards (totalRewards)
+            totalRewards += rewardWei;
+        }
+        // Todo: emit event
+        return true;
+    }
+
+
+    function claimStakerReward(address staker, uint percentageinPrecision, uint epoch) public {
+        // onlyDAO
+        // send reward
+        // update rewardPerEpoch
+        // update totalReward
+    }
+
+    function claimReserveRebate(address reserve) public {
+        // only DAO
+        // send rebate to reserve
+        // update rebatePerReserve;
+        // update total rebate amounts?
+        // update reserve rebate to 1 (avoid 0...) otherwise div by 0 issue? but will we even need to div by 0?
+        // if we include a dest address, we need an owner / admin of the reserve and the below function.
+    }
+
+    function burnKNC() public {
+        // only DAO?
+        // convert fees to KNC and burn
+        // Eth for burning is the remaining == (total balance - total_reward_amount - total_reserve_rebate).
+    }
+
+    function setReserveRebateWallet(address reserve, address wallet) public returns (bool){}
```

Comment:
we will not set rebate wallet here.
we receive an array of rebate wallets and do accounting.
since we don't have admin to set rebate wallets.

Reply:
Right, maybe it's more appropriate to scope it within KyberNetwork.sol? I'm just thinking about the case where the reserveRebateWallet was not set when the reserve is added or if they decided to change it in the future. 

Also, don't worry i'm not spending much time on it that's why I haven't implemented it!

Reply:
the fee handler will get a list of rebate wallets. and update their accounts with internal accounting.

if the reserve updates a wallet in the future. fee handler doesn't need to know about it.
any one who updated his wallet will have to know he should claim for different wallets.

Reply:
noted